pkg/storegateway/indexheader/reader_pool.go

```diff
@@ -88,26 +92,49 @@ func (l lazyLoadedHeadersSnapshot) persist(persistDir string) error {
 
 // NewReaderPool makes a new ReaderPool. If lazy-loading is enabled, NewReaderPool also starts a background task for unloading idle Readers and persisting a list of loaded Readers to disk.
 func NewReaderPool(logger log.Logger, lazyReaderEnabled bool, lazyReaderIdleTimeout time.Duration, metrics *ReaderPoolMetrics, lazyLoadedSnapshotConfig LazyLoadedHeadersSnapshotConfig) *ReaderPool {
-	p := newReaderPool(logger, lazyReaderEnabled, lazyReaderIdleTimeout, metrics)
+	var snapshot *lazyLoadedHeadersSnapshot
+	if lazyReaderEnabled && lazyLoadedSnapshotConfig.EagerLoadingEnabled {
+		lazyLoadedSnapshotFileName := filepath.Join(lazyLoadedSnapshotConfig.Path, lazyLoadedHeadersListFile)
+		var loadSnapshotErr error
+		snapshot, loadSnapshotErr = loadLazyLoadedHeadersSnapshot(lazyLoadedSnapshotFileName)
+		if loadSnapshotErr != nil {
+			level.Warn(logger).Log("msg", "loading the list of index-headers from file failed; starting with no loaded index-headers", "file", lazyLoadedSnapshotFileName, "err", err)
+		}
+		// We will remove the file regardless the value of loadSnapshotErr.
+		// In the case such as snapshot loading causing OOM, we will still
+		// remove the snapshot file and do lazyLoading after server is restarted.
+		err := os.Remove(lazyLoadedSnapshotFileName)
+		if err != nil {
+			level.Warn(logger).Log("msg", "removing the list of index-headers file failed", "file", lazyLoadedSnapshotFileName, "err", err)
```

Comment:
[nit] Given we use the term "snapshot" internally to refer to the list of index-headers, should we use the same term in this message and the one above?